src/board.rs

```diff
@@ -0,0 +1,1021 @@
+            return Err("Every player needs a root and orientation");
+        }
+
+        // Transform string into a board
+        let mut squares = vec![];
+        for line in s.split('\n') {
+            let mut squares_in_line: Vec<Option<Square>> = vec![];
+            for (i, letter) in line.chars().enumerate() {
+                if i % 2 == 1 {
+                    if letter != ' ' {
+                        return Err("board strings should have spaces to separate each tile");
+                    }
+                } else if letter == ' ' {
+                    squares_in_line.push(None);
+                } else if letter == '_' {
+                    squares_in_line.push(Some(Square::Empty));
+                } else {
+                    squares_in_line.push(Some(Square::Occupied(0, letter)));
+                }
+            }
+            squares.push(squares_in_line);
+        }
+
+        // Make sure the board is an valid non-jagged grid
+        for line in squares.iter().skip(1) {
+            if line.len() != squares[0].len() {
+                return Err("Unequal line lengths");
+            }
+        }
+
+        // Make sure letters connected to players' roots are owned by the player
+        let r = roots.clone(); // TODO: remove hack
+        let mut board = Self {
+            roots,
+            squares,
+            orientations,
+        };
+        for (player, root) in r.iter().enumerate() {
+            if player != 0 {
+                // All tiles are already owned by the first player by default
+                for square in board.depth_first_search(*root).iter() {
+                    if let Ok(Square::Occupied(_, value)) = board.get(*square) {
+                        board.set(*square, player, value).expect(
+                            "A coordinate returned from a DFS should always be valid and settable",
+                        );
+                    }
+                }
+            }
+        }
+
+        Ok(board)
+    }
+
+    // TODO: generic board constructor that accepts a grid of squares with arbitrary values, as long as:
+    //  - the empty squares are fully connected
+    //  - there are at least 2 roots
+    //  - the roots are at empty squares
+
+    pub fn get(&self, position: Coordinate) -> Result<Square, &str> {
+        if position.y < 0 || position.x < 0 {
+            return Err("negative coordinates");
+        };
+        let x = position.x as usize;
+        let y = position.y as usize;
+
+        if y >= self.squares.len() {
+            Err("y-coordinate is too large for board height") // TODO: specify the coordinate and height
+        } else if x >= self.squares[0].len() {
+            Err("x-coordinate is too large for board width") // TODO: specify the coordinate and width
+        } else {
+            match self.squares[y][x] {
+                None => Err("Invalid position"),
+                Some(square) => Ok(square),
+            }
+        }
```

Comment:
I changed set and get. I think it's harder to read the error logic, but arguably easier to read the happy path. Can you think of a way to make it more succinct? I don't think I can use `?` since I want to return a `Result` whereas `get` and `get_mut` return `Option`.

Reply:
Have you pushed that? Not seeing it

Reply:
Oops pushed now

Reply:
Opened a PR with my suggestion:

https://github.com/BlakeMScurr/Truncate/pull/4/files